### PR TITLE
TFX Iris sample

### PR DESCRIPTION
**Pull request by numerology**
Illustrative purpose

~DO NOT SUBMIT~ Changed my mind. I think we should include this sample :)

Verified at https://3ed47013582d1554-dot-us-central2.pipelines.googleusercontent.com/#/runs/details/a17f06a2-7799-4ad0-9bde-bc9c275b6355

Environment:
SDK: TFX 0.21.0 + KFP 0.2.4
Runtime: Hosted pipeline 0.2.3

- [x] Sample test coverage
- [x] Prebuilt sample config

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/3119)

### CODE REVIEW

**Comment by numerology** [comment]
/cc @jingzhang36 
FYI

File: samples/core/iris/iris_pipeline.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+if __name__ == '__main__':
+  absl.logging.set_verbosity(absl.logging.INFO)
+  # Make sure the version of TFX image used is consistent with the version of
+  # TFX SDK. Here we use tfx:0.15.0 image.
+  config = kubeflow_dag_runner.KubeflowDagRunnerConfig(
+      kubeflow_metadata_config=kubeflow_dag_runner.
+      get_default_kubeflow_metadata_config(),
+      pipeline_operator_funcs=list(
+          filter(
+              lambda operator: operator.__name__.find('gcp_secret') == -1,
```
Comment (Ark-kun) [inline samples/core/iris/iris_pipeline.py]: Have we removed auto-applying `use_gcp_secret` in TFX master?

Comment (numerology) [reply samples/core/iris/iris_pipeline.py]: We did in 0.21.0 formal release

I'm doing version bumping in another PR.

Comment (numerology) [reply samples/core/iris/iris_pipeline.py]: Done

File: samples/core/iris/iris_pipeline.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+_module_file_param = data_types.RuntimeParameter(
+    name='module-file',
+    default='gs://ml-pipeline-playground/iris/modules/iris_utils_native_keras.py',
+    ptype=Text,
+)
+
+# Directory and data locations. This example assumes all of the flowers
+# example code and metadata library is relative to a GCS path.
+_pipeline_root = os.path.join(
+    'gs://jxzheng-helloworld', 'tfx_iris', kfp.dsl.RUN_ID_PLACEHOLDER
```
Comment (Ark-kun) [inline samples/core/iris/iris_pipeline.py]: Can we use the "gs://<your-gcs-bucket>/" placeholder here?

File: samples/core/iris/iris_pipeline.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+from tfx.components import CsvExampleGen
+from tfx.components import Evaluator
+from tfx.components import ExampleValidator
+from tfx.components import SchemaGen
+from tfx.components import StatisticsGen
+from tfx.components import Trainer
+from tfx.components import Transform
+from tfx.components.base import executor_spec
+from tfx.components.trainer.executor import GenericExecutor
```
Comment (Ark-kun) [inline samples/core/iris/iris_pipeline.py]: I wonder what would be the best option when we want to import executors from multiple components.

Comment (numerology) [reply samples/core/iris/iris_pipeline.py]: Ummm. That's a good question. I don't see any shortcut to that.

I guess partly that's because we don't often explicitly interact with TFX component executor when authoring the pipeline.

**Comment by jingzhang36** [comment]
I have a question regarding the Evaluator output of this pipeline. I did a run of this pipeline and the Evaluator's output in my public GCS bucket gs://jingzhangjz-project-outputs/tfx_iris/8d8d58b1-2826-4533-ae0d-c04a65b2864d/Evaluator/evaluation/30.

While I used the following to parse the metric file under that bucket

import tensorflow as tf 
from tensorflow_model_analysis.proto import metrics_for_slice_pb2
from typing import Any
for record in tf.compat.v1.python_io.tf_record_iterator('gs://jingzhangjz-project-outputs/tfx_iris/8d8d58b1-2826-4533-ae0d-c04a65b2864d/Evaluator/evaluation/30/metrics'):
    metrics_for_slice = metrics_for_slice_pb2.MetricsForSlice.FromString(record)
    print(metrics_for_slice)

The parse data seems missing metrics field. They are like 
slice_key {
  single_slice_keys {
    column: "sepal_length"
    float_value: 4.900000095367432
  }
}
slice_key {
  single_slice_keys {
    column: "sepal_length"
    float_value: 5.699999809265137
  }
}
slice_key {
...
}
....

There are all slice_key fields, but no metrics field. As a reference, the taxi example in KFP has  parsed data that looks like:
slice_key{
...
}
metrics {
  key: "precision"
  value {
    double_value {
    }
  }
}
metrics {
}

**Comment by numerology** [comment]
> There are all slice_key fields, but no metrics field.

It seems like a Keras problem IIUC. Switching to vanilla TF solved this.

File: samples/core/iris/iris_pipeline.py (hunk 2 above)
Comment (numerology) [reply samples/core/iris/iris_pipeline.py]: Will update this PR after today's bugbash. I want to make sure that magic works :)

**Comment by numerology** [comment]
> > There are all slice_key fields, but no metrics field.
> 
> It seems like a Keras problem IIUC. Switching to vanilla TF solved this.

And if you indeed want to use Keras, you'll need to handcraft the `eval_config` like following:

```python
eval_config = tfma.EvalConfig(
      model_specs=[
          tfma.ModelSpec(name='candidate', label_key='variety'),
          tfma.ModelSpec(
              name='baseline', label_key='variety', is_baseline=True)
      ],
      slicing_specs=[tfma.SlicingSpec()],
      metrics_specs=[
          tfma.MetricsSpec(metrics=[
              tfma.MetricConfig(
                  class_name='SparseCategoricalAccuracy',
                  threshold=tfma.config.MetricThreshold(
                      value_threshold=tfma.GenericValueThreshold(
                          lower_bound={'value': 0.9}),
                      change_threshold=tfma.GenericChangeThreshold(
                          direction=tfma.MetricDirection.HIGHER_IS_BETTER,
                          absolute={'value': -1e-10})))
          ])])
```

File: samples/core/iris/iris_pipeline.py (hunk 2 above)
Comment (numerology) [reply samples/core/iris/iris_pipeline.py]: Okay it only partially works :- ( 

Fixed the backend in #3145

**Comment by numerology** [comment]
@rmgogogo , this is the forked native Keras sample.

File: backend/src/apiserver/config/sample_config.json [hunk 4]
```diff
@@ -18,5 +18,10 @@
     "name": "[Tutorial] DSL - Control structures",
     "description": "[source code](https://github.com/kubeflow/pipelines/tree/master/samples/tutorials/DSL%20-%20Control%20structures) Shows how to use conditional execution and exit handlers. This pipeline will randomly fail to demonstrate that the exit handler gets executed even in case of failure.",
     "file": "/samples/tutorials/DSL - Control structures/DSL - Control structures.py.yaml"
+  },
+  {
+    "name": "[Demo] TFX - Iris classification pipeline",
+    "description": "[source code](https://github.com/kubeflow/pipelines/tree/master/samples/core/iris). Example pipeline that classifies Iris flower subspecies and how to use native Keras within TFX.",
```
Comment (rmgogogo) [inline backend/src/apiserver/config/sample_config.json]: Just FYI, after sync with Yuan, I removed the magic in FE Dockerfile which replace the "master" when build a docker images. So here the link is hard coded.

File: samples/core/iris/iris.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+      model_blessing=model_analyzer.outputs['blessing'],
+      push_destination=pusher_pb2.PushDestination(
+          filesystem=pusher_pb2.PushDestination.Filesystem(
+              base_directory=os.path.
+              join(str(pipeline.ROOT_PARAMETER), 'model_serving')
+          )
+      )
+  )
+
+  # TODO(numerology): Add pusher and model validator.
```
Comment (rmgogogo) [inline samples/core/iris/iris.py]: refine this todo? you already have pusher.

as for model validator, remember you discussed native keras will replace it with another one? I may forget or remember it in a wrong direction.

File: samples/core/iris/iris.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+
+
+if __name__ == '__main__':
+  absl.logging.set_verbosity(absl.logging.INFO)
+  # Make sure the version of TFX image used is consistent with the version of
+  # TFX SDK. Here we use tfx:0.21.1 image.
+  config = kubeflow_dag_runner.KubeflowDagRunnerConfig(
+      kubeflow_metadata_config=kubeflow_dag_runner.
+      get_default_kubeflow_metadata_config(),
+      tfx_image='tensorflow/tfx:0.21.2',
```
Comment (rmgogogo) [inline samples/core/iris/iris.py]: OK we retag it later, just the topic of docker.io

File: samples/core/iris/iris.py (hunk 5 above)
Comment (numerology) [reply samples/core/iris/iris.py]: Are sorry I forgot to clean this up. Done.

File: samples/core/iris/iris.py (hunk 6 above)
Comment (numerology) [reply samples/core/iris/iris.py]: Ack. Will do that together with other prebuilt sample.
